### DISCUSSION TITLE
Error during inference on multi-speaker model

### DISCUSSION THREAD

**Discussion by astricks** [coqui-ai/TTS#553] [General Q&A] [answered]
I have trained a multi-speaker model and am trying to do inference on said model. However, I am hitting an issue with an 
"unexpected key".

	**Unexpected key(s) in state_dict: "speaker_embedding.weight".** 

I am not using an external speaker embeddings file.

The console logs are below
-->8--
(tts) ~/code/TTS$ tts  --text "मैं बता रहा हूँ, मुझे यहाँ कुछ ठीक नहीं लग रहा, यार।" --model_path ../output/hindi-May-22-2021_08+18PM-56ae83a/best_model.pth.tar --config_path ../output/hindi-May-22-2021_08+18PM-56ae83a/config.json --out_path /tmp/angryspeaker.wav --vocoder_name "vocoder_models/universal/libri-tts/wavegrad" --speaker_idx 0 --speakers_file_path ../output/hindi-May-22-2021_08+18PM-56ae83a/speakers.json

 > vocoder_models/universal/libri-tts/wavegrad is already downloaded.
Loading speakers ...
speaker_id_file_path:../output/hindi-May-22-2021_08+18PM-56ae83a/speakers.json
 > Using model: Tacotron2
Traceback (most recent call last):
  File "/home/arvind/.venv/tts/bin/tts", line 11, in <module>
    sys.exit(main())
  File "/home/arvind/code/TTS/TTS/bin/synthesize.py", line 234, in main
    args.use_cuda,
  File "/home/arvind/code/TTS/TTS/utils/synthesizer.py", line 72, in __init__
    self._load_tts(tts_checkpoint, tts_config_path, use_cuda)
  File "/home/arvind/code/TTS/TTS/utils/synthesizer.py", line 142, in _load_tts
    self.tts_model.load_checkpoint(self.tts_config, tts_checkpoint, eval=True)
  File "/home/arvind/code/TTS/TTS/tts/models/tacotron_abstract.py", line 132, in load_checkpoint
    self.load_state_dict(state["model"])
  File "/home/arvind/.venv/tts/lib/python3.6/site-packages/torch/nn/modules/module.py", line 1224, in load_state_dict
    self.__class__.__name__, "\n\t".join(error_msgs)))
RuntimeError: Error(s) in loading state_dict for Tacotron2:
	Unexpected key(s) in state_dict: "speaker_embedding.weight". 
	size mismatch for decoder.attention_rnn.weight_ih: copying a param with shape torch.Size([4096, 1280]) from checkpoint, the shape in current model is torch.Size([4096, 768]).
	size mismatch for decoder.attention.inputs_layer.linear_layer.weight: copying a param with shape torch.Size([128, 1024]) from checkpoint, the shape in current model is torch.Size([128, 512]).
	size mismatch for decoder.decoder_rnn.weight_ih: copying a param with shape torch.Size([4096, 2048]) from checkpoint, the shape in current model is torch.Size([4096, 1536]).
	size mismatch for decoder.linear_projection.linear_layer.weight: copying a param with shape torch.Size([320, 2048]) from checkpoint, the shape in current model is torch.Size([320, 1536]).
-->8--

The config file for the training is below.
-->8--
{
"github_branch":"* hindi0.0.13",
    "model": "Tacotron2",
    "run_name": "hindi",
    "run_description": "hindi without gst",

    // AUDIO PARAMETERS
    "audio":{
        // stft parameters
        "fft_size": 1024,         // number of stft frequency levels. Size of the linear spectogram frame.
        "win_length": 1024,      // stft window length in ms.
        "hop_length": 256,       // stft window hop-lengh in ms.
        "frame_length_ms": null, // stft window length in ms.If null, 'win_length' is used.
        "frame_shift_ms": null,  // stft window hop-lengh in ms. If null, 'hop_length' is used.

        // Audio processing parameters
        "sample_rate": 22050,   // DATASET-RELATED: wav sample-rate.
        "preemphasis": 0.0,     // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.
        "ref_level_db": 20,     // reference level db, theoretically 20db is the sound of air.

        // Silence trimming
        "do_trim_silence": true,// enable trimming of slience of audio as you load it. LJspeech (true), TWEB (false), Nancy (true)
        "trim_db": 40,          // threshold for timming silence. Set this according to your dataset.

        // Griffin-Lim
        "power": 1.5,           // value to sharpen wav signals after GL algorithm.
        "griffin_lim_iters": 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.

        // MelSpectrogram parameters
        "num_mels": 80,         // size of the mel spec frame.
        "mel_fmin": 50.0,        // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!
        "mel_fmax": 10000.0,     // maximum freq level for mel-spec. Tune for dataset!!
        "spec_gain": 1,

        // Normalization parameters
        "signal_norm": true,    // normalize spec values. Mean-Var normalization if 'stats_path' is defined otherwise range normalization defined by the other params.
        "min_level_db": -100,   // lower bound for normalization
        "symmetric_norm": true, // move normalization to range [-1, 1]
        "max_norm": 4.0,        // scale normalization to range [-max_norm, max_norm] or [0, max_norm]
        "clip_norm": true,      // clip normalized values into the range.
        "stats_path": null      // DO NOT USE WITH MULTI_SPEAKER MODEL. scaler stats file computed by 'compute_statistics.py'. If it is defined, mean-std based notmalization is used and other normalization params are ignored
    },

    // VOCABULARY PARAMETERS
    // if custom character set is not defined,
    // default set in symbols.py is used
    "characters":{
        "pad": "_",
        "eos": "~",
        "bos": "^",
        "characters": "अआइईउऊऋएऐऑओऔकखगघङचछजझञटठडढणतथदधनपफबभमयरलवशषसहह़ािीुूृेैॉोौ्",
        "punctuations":"!,-.? ",
        "phonemes":"iyɨʉɯuɪʏʊeøɘəɵɤoɛœɜɞʌɔæɐaɶɑɒᵻʘɓǀɗǃʄǂɠǁʛpbtdʈɖcɟkɡqɢʔɴŋɲɳnɱmʙrʀⱱɾɽɸβfvθðszʃʒʂʐçʝxɣχʁħʕhɦɬɮʋɹɻjɰlɭʎʟˈˌːˑʍwɥʜʢʡɕʑɺɧɚ˞ɫ"
    },

    // DISTRIBUTED TRAINING
    "distributed":{
        "backend": "nccl",
        "url": "tcp:\/\/localhost:54321"
    },

    "reinit_layers": [],    // give a list of layer names to restore from the given checkpoint. If not defined, it reloads all heuristically matching layers.

    // TRAINING
    "batch_size": 16,       // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.
    "eval_batch_size":16,
    "r": 4,                 // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled.
    "gradual_training": null, 
    "mixed_precision": false,     // level of optimization with NVIDIA's apex feature for automatic mixed FP16/FP32 precision (AMP), NOTE: currently only O1 is supported, and use "O1" to activate.

    // LOSS SETTINGS
    "loss_masking": true,       // enable / disable loss masking against the sequence padding.
    "decoder_loss_alpha": 0.5,  // original decoder loss weight. If > 0, it is enabled
    "postnet_loss_alpha": 0.25, // original postnet loss weight. If > 0, it is enabled
    "postnet_diff_spec_alpha": 0.25,     // differential spectral loss weight. If > 0, it is enabled
    "decoder_diff_spec_alpha": 0.25,     // differential spectral loss weight. If > 0, it is enabled
    "decoder_ssim_alpha": 0.5,     // decoder ssim loss weight. If > 0, it is enabled
    "postnet_ssim_alpha": 0.25,     // postnet ssim loss weight. If > 0, it is enabled
    "ga_alpha": 5.0,           // weight for guided attention loss. If > 0, guided attention is enabled.
    "stopnet_pos_weight": 15.0, // pos class weight for stopnet loss since there are way more negative samples than positive samples.

    // VALIDATION
    "run_eval": true,
    "test_delay_epochs": 10,  //Until attention is aligned, testing only wastes computation time.
    "test_sentences_file": "/home/arvind/data/hindi/metadata_test_sentences.txt",  // set a file to load sentences to be used for testing. If it is null then we use default english sentences.

    // OPTIMIZER
    "noam_schedule": false,        // use noam warmup and lr schedule.
    "grad_clip": 1.0,              // upper limit for gradients for clipping.
    "epochs": 20000,               // total number of epochs to train.
    "lr": 0.0001,                  // Initial learning rate. If Noam decay is active, maximum learning rate.
    "wd": 0.000001,                // Weight decay weight.
    "warmup_steps": 4000,          // Noam decay steps to increase the learning rate from 0 to "lr"
    "seq_len_norm": false,         // Normalize eash sample loss with its length to alleviate imbalanced datasets. Use it if your dataset is small or has skewed distribution of sequence lengths.

    // TACOTRON PRENET
    "memory_size": -1,             // ONLY TACOTRON - size of the memory queue used fro storing last decoder predictions for auto-regression. If < 0, memory queue is disabled and decoder only uses the last prediction frame.
    "prenet_type": "original",     // "original" or "bn".
    "prenet_dropout": true,       // enable/disable dropout at prenet.

    // TACOTRON ATTENTION
    "attention_type": "original",  // 'original' , 'graves', 'dynamic_convolution'
    "attention_heads": 4,          // number of attention heads (only for 'graves')
    "attention_norm": "sigmoid",   // softmax or sigmoid.
    "windowing": false,            // Enables attention windowing. Used only in eval mode.
    "use_forward_attn": false,     // if it uses forward attention. In general, it aligns faster.
    "forward_attn_mask": false,    // Additional masking forcing monotonicity only in eval mode.
    "transition_agent": false,     // enable/disable transition agent of forward attention.
    "location_attn": true,         // enable_disable location sensitive attention. It is enabled for TACOTRON by default.
    "bidirectional_decoder": false,  // use https://arxiv.org/abs/1907.09006. Use it, if attention does not work well with your dataset.
    "double_decoder_consistency": false,  // use DDC explained here https://erogol.com/solving-attention-problems-of-tts-models-with-double-decoder-consistency-draft/
    "ddc_r": 4,                           // reduction rate for coarse decoder.

    // STOPNET
    "stopnet": true,               // Train stopnet predicting the end of synthesis.
    "separate_stopnet": true,      // Train stopnet seperately if 'stopnet==true'. It prevents stopnet loss to influence the rest of the model. It causes a better model, but it trains SLOWER.

    // TENSORBOARD and LOGGING
    "print_step": 25,       // Number of steps to log training on console.
    "tb_plot_step": 100,    // Number of steps to plot TB training figures.
    "print_eval": true,     // If True, it prints intermediate loss values in evalulation.
    "save_step": 5000,      // Number of training steps expected to save traninpg stats and checkpoints.
    "checkpoint": true,     // If true, it saves checkpoints per "save_step"
    "keep_all_best": false,  // If true, keeps all best_models after keep_after steps
    "keep_after": 10000,    // Global step after which to keep best models if keep_all_best is true
    "tb_model_param_stats": false,     // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.

    // DATA LOADING
    "text_cleaner": "basic_cleaners",
    "enable_eos_bos_chars": false, // enable/disable beginning of sentence and end of sentence chars.
    "num_loader_workers": 2,        // number of training data loader processes. Don't set it too big. 4-8 are good values.
    "num_val_loader_workers": 2,    // number of evaluation data loader processes.
    "batch_group_size": 32,  //Number of batches to shuffle after bucketing.
    "min_seq_len": 24,      // DATASET-RELATED: minimum text length to use in training
    "max_seq_len": 130,     // DATASET-RELATED: maximum text length
    "compute_input_seq_cache": false,  // if true, text sequences are computed before starting training. If phonemes are enabled, they are also computed at this stage.
    "use_noise_augment": true,

    // PATHS
    "output_path": "/home/arvind/code/output",

    // PHONEMES
    "phoneme_cache_path": "/home/arvind/code/cache/",  // phoneme computation is slow, therefore, it caches results in the given folder.
    "use_phonemes": false,           // use phonemes instead of raw characters. It is suggested for better pronounciation.
    "phoneme_language": "hi",     // depending on your target language, pick one from  https://github.com/bootphon/phonemizer#languages

    // MULTI-SPEAKER and GST
    "use_speaker_embedding": true,      // use speaker embedding to enable multi-speaker learning.
    "use_gst": false,          	        // use global style tokens
    "use_external_speaker_embedding_file": false, // if true, forces the model to use external embedding per sample instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "external_speaker_embedding_file": null, // if not null and use_external_speaker_embedding_file is true, it is used to load a specific embedding file and thus uses these embeddings instead of nn.embeddings, that is, it supports external embeddings such as those used at: https://arxiv.org/abs /1806.04558
    "gst":	{			                // gst parameter if gst is enabled
        "gst_style_input": null,        // Condition the style input either on a
                                        // -> wave file [path to wave] or
                                        // -> dictionary using the style tokens {'token1': 'value', 'token2': 'value'} example {"0": 0.15, "1": 0.15, "5": -0.15}
                                        // with the dictionary being len(dict) <= len(gst_style_tokens).
        "gst_embedding_dim": 512,
        "gst_num_heads": 4,
        "gst_style_tokens": 10,
        "gst_use_speaker_embedding": false
	},

    // DATASETS
    "datasets":   // List of datasets. They all merged and they get different speaker_ids.
        [
            {
                "name": "hindi",
                "path": "/home/arvind/data/hindi",
                "meta_file_train": "metadata_train.txt", // for vtck if list, ignore speakers id in list for train, its useful for test cloning with new speakers
                "meta_file_val": null 
            }
        ]
}
-->8--

The generated speakers file is below.
-->8--
{
       "Angry": 0,
       "Happy": 1,
       "Neutral": 2,
       "Sad": 3,
       "Scared": 4,
       "Surprised": 5,
       "Voiceover": 6
       
}
-->8--

I'd appreciate any help with debugging what I'm doing wrong. I have been playing with the code but unable to figure out why this issue persists.

**Comment by lexkoro** [comment, marked as answer]
which commit are you using?

**Reply by astricks** [reply]
release 0.0.13

**Reply by lexkoro** [reply]
I think there is actually a bug in the synthesizer. I'll take a look and get back to you later.

**Reply by lexkoro** [reply]
I didn't get the error you are having but still had to make some adjustments for it to work.
See here:
https://github.com/lexkoro/TTS-test/commit/9f93b2af31902a1dd5d3fdee15103ec7d530a58e#diff-2e7fe81d1ddfbbdd3f7534fdf4335deb4b7c2a020f22ea4f22df0609f9155b37

**Reply by astricks** [reply]
I re-installed TTS version 0.0.13, made these adjustments and it worked! Thank you!
